### ISSUE TITLE
Improve the performance of vim macros

### ISSUE THREAD

**Issue by dev-ardi** [neovim/neovim#24920]
### Problem

Macros can be slow when repeated a big number of times. 

### Expected behavior

First, I'd be willing to implement this myself with some guidance.

I don't know how macros work under the hood. If they're implemented in the naive way, to just repeat the commands input by the user there might be some ways to optimize them:

1. Eliminate no-ops. For example, repeated esc, j followed by a k, etc.
2. A macro compiler that turns macros into the native movements or something like that. Resolve what the movements will be ahead of time instead of having to pass through the whole input decoding and mapping (I've no idea how vim works)

**Comment by matveyt** [comment]
If you talk of mappings then, yes, they are implemented by manipulating an input queue. But I wouldn't call it "naive" but rather "true" way. That allows to track all possible changes in bookmarks, jump list, change list, undo list etc.etc.

Of course, there is some price in performance to pay. However, mappings are not THAT slow to avoid them completely. At least, try to do `set lazyredraw` first, as the screen redraw tends to be the most time consuming part by margin. If it still looks slow then switch to scripting. Even legacy VimL should have sufficient performance in 99.9% cases.

On the other hand, after "optimization" the mappings would inevitably lose their "true evaluation" feature. And then an enormous heap of incompatibilities and bugs would appear.

**Comment by lewis6991** [comment]
I agree with @matveyt , this is certainly non-trivial, and probably not realistically feasible.

I think unless there is a concrete proposal, there isn't a way to progress this.

**Comment by dev-ardi** [comment]
> Of course, there is some price in performance to pay. However, mappings are not THAT slow. At least, try to do `set lazyredraw` first, as the screen redraw tends to be the most time consuming part by margin. Yet another typical problem is the misuse of `set clipboard=unnamedplus` option that may severely affect the performance. Just avoid it.

Maybe only copy to clipboard on the last part, or do it asynchronously?

**Comment by matveyt** [comment]
> Maybe only copy to clipboard on the last part

Yes, and this why setting the `clipboard` option is useless.

**Comment by dev-ardi** [comment]
How is it useless? I don't want to do the whole thing to copy and paste to system clipboard. I think most people use that feature

**Comment by matveyt** [comment]
@dev-ardi Do you know what `clipboard=unnamedplus` does _precisely_?

**Comment by matveyt** [comment]
This is not the internals. This is user option that you think "most people use". It makes to copy to system clipboard _by default_ like in other editors. While normal Vim default is to copy to internal register unless clipboard is required explicitly.

**Comment by dev-ardi** [comment]
I'm missing your point here, you're saying that it's slow?

**Comment by matveyt** [comment]
There is nothing good in it but it _may_ give some problems including slowing down mappings and/or scripts.

**Comment by justinmk** [comment]
Related:

- https://github.com/neovim/neovim/issues/3508

**Comment by konosubakonoakua** [comment]
```lua
-- NOTE: donot trigger autocmd when executing macro
-- https://www.reddit.com/r/neovim/comments/tsol2n/comment/i2ugipm/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button
vim.cmd([[
  xnoremap @ :<C-U>execute "noautocmd '<,'>norm! " . v:count1 . "@" . getcharstr()<cr>
  nnoremap @ <cmd>execute "noautocmd norm! " . v:count1 . "@" . getcharstr()<cr>
]])

This helps me a lot.
```

**Comment by glav-git** [comment]
@oriongonza It would be nice to clarify, do you mean "raw" performance of macros in nvim without any configuration or are we talking about the performance of macros with syntax highlighting enabled (for example treesitter), lsp and additional plugins in general?

I ask because I recently learned that plugins can greatly affect the speed of macro execution. And all because, as I understand, tools like treesitter and lsp actually work with every change in the text in most cases, and as you can imagine, when executing a macro, there are a lot of such changes.

In my case, I found this great plugin: https://github.com/pteroctopus/faster.nvim 
It disables all these additional features of the editor for the duration of the macro execution, which in my case increased the speed of macro execution to more than acceptable speed!

If you are not sure whether your problem is related to plugins or not, you can test it as follows: try opening a file with the following command `nvim -u none filename` and execute your "slow macro", comparing it speed execution with the speed you usually experience. If you notice an improvement, I think faster.nvim is what you need.

Also, if it's really the case and lsp can slow down macro execution, since lsp is built-in feature of nvim i guess we can make argument, that at least in some cases like this nvim could try implement this optimization, i guess.
